docs/develop/cmodules.rst

```diff
@@ -0,0 +1,161 @@
+MicroPython external C modules
+==============================
+
+When developing modules for use with micropython you may find you run into
+limitations with the python environment, often due to an inability to access
+certain hardware resources or python speed limitations.
+
+If your limitations can't be resolved with suggestions in :ref:`speed_python`,
+writing some or all of your module in C is a viable option.
+
+If your module is designed to access or work with commonly available 
+hardware or libraries please consider implementing it inside the micropython
+source tree alongside similar modules and submitting it as a pull request. 
+If however you're targeting obscure or proprietary systems it may make 
+more sense to keep this external to the main micropython repository. 
+
+This chapter describes how to compile such external modules into the 
+micropython executable or firmware image.
+
+
+Structure of an external C module
+---------------------------------
+
+A micropython user c module is a directory with the following files:
+
+* ``*.c`` and/or ``*.h`` source code files for your module.
+
+  These will typically include the low level functionality being implemented and 
+  the micropython binding functions to expose the functions and module(s). 
+
+  Currently the best reference for writing these functions/modules is 
+  to find similar modules within the micropython tree and use them as examples.
+
+* ``micropython.mk`` contains the Makefile fragment for this module.
+
+  ``$(USERMOD_DIR)`` is available in ``micropython.mk`` as the path to your
```

Comment:
Just an idea, but wouldn't it be better to more clearly state the intent and have this called PY_USERMOD_DIR ?

Reply:
Personally I think adding `PY_` to it makes it sound like a folder you put python files in, rather than c files.